### PR TITLE
Define the term data plane volatile and how it relates to read-write symmetry

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

**Comment by jafingerhut** [comment]
Intended to resolve https://github.com/p4lang/p4runtime/issues/408

**Comment by jafingerhut** [comment]
TODO: Does the MessageDifferencer function mentioned in the spec ignore all of the data plane volatile fields when comparing read values against write values, at least for things that are data plane volatile in PSA?

I would guess it would need to be updated for the new things that can be data plane volatile in PNA.

**Comment by jafingerhut** [comment]
@smolkaj @jonathan-dilorenzo @verios-google FYI please read it over to see if you have any corrections/comments/questions.

**Comment by smolkaj** [comment]
@verios-google for visibility

**Review by verios-google — COMMENTED** [review]
Great change, good job including the PSA and PNA specific items.
I have a question, would the field `time_since_last_hit` (Protobuf with `elapsed_ns`) still be considered data plane volatile in non-PSA/non-PNA implementations? Or would that depend on the implementation?

**Comment by jafingerhut** [comment]
> Great change, good job including the PSA and PNA specific items. I have a question, would the field `time_since_last_hit` (Protobuf with `elapsed_ns`) still be considered data plane volatile in non-PSA/non-PNA implementations? Or would that depend on the implementation?

It would depend upon the behavior of that architecture, but if you are asking about the v1model architecture, then yes it is data plane volatile.

I have a difficult time imagining an implementation that claims to support idle timeout in a way compatible with the existing defined P4Runtime API messages, including their documentation, where those fields were not data plane volatile.  But it seems safest to qualify the documentation with the architectures we know, rather than ones that we do not.

File: docs/v1/P4Runtime-Spec.mdk [hunk 1]
```diff
@@ -2185,6 +2185,114 @@ C++ API supports comparing messages while treating repeated fields as sets, so
 that different orderings of the same elements are considered equal. This method
 of comparing Protobuf messages may come at a cost in performance.
 
+### Data plane volatile objects
+
+An exception to read-write symmetry are objects whose contents or
```
Comment (smolkaj) [inline docs/v1/P4Runtime-Spec.mdk:2190]: Nit: Should we briefly mention this exception and add a forward reference to this section at the beginning of the previous section, so people don't miss it?

**Review by smolkaj — APPROVED** [review]
This looks great from my PoV. Thanks Andy!

File: docs/v1/P4Runtime-Spec.mdk (hunk 1 above)
Comment (jafingerhut) [reply docs/v1/P4Runtime-Spec.mdk:2192]: Done

**Review by antoninbas — APPROVED** [review]
sorry for the delay, LGTM

I intend to merge this by end of day

File: docs/v1/P4Runtime-Spec.mdk [hunk 2]
```diff
@@ -2185,6 +2187,114 @@ C++ API supports comparing messages while treating repeated fields as sets, so
...
+field of a `TableEntry` can be modified by the data plane when packets
+match the entry.
+
+For a table with a direct meter associated with it, the
+`meter_counter_data` field of a `TableEntry` can be modified by the data
+plane when packets match the entry.
+
+For a PSA [@PSA] table with property `psa_idle_timeout` equal to
+`PSA_IdleTimeout_t.NOTIFY_CONTROL`, the data plane can modify the
+`elapsed_ns` field of a `TabeEntry` when _no_ packets match the entry
```
Comment (jonathan-dilorenzo) [inline docs/v1/P4Runtime-Spec.mdk:2220]: nit: TabeEntry -> TableEntry

**Review by jonathan-dilorenzo — APPROVED** [review]
Super nice! Thanks Andy!

File: docs/v1/P4Runtime-Spec.mdk (hunk 2 above)
Comment (jafingerhut) [reply docs/v1/P4Runtime-Spec.mdk]: fixed

**Pull request merged**